tensorflow_datasets/core/custom_dataset/image_folder.py

```diff
@@ -0,0 +1,176 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Lint as: python3
+"""Image Classification Folder datasets."""
+
+import os
+from typing import Dict, List, Tuple, Set
+
+import tensorflow.compat.v2 as tf
+from tensorflow_datasets import core
+
+# Dict of `split_label_images['split']['Label']` = `List['images']`
+SplitLabelImageDict = Dict[str, Dict[str, List[str]]]
+
+_SUPPORTED_IMAGE_FORMAT = ('.jpg', '.jpeg', '.png')
+
+
+class ImageFolder(core.DatasetBuilder):
+  """Generic image classification dataset created from manual directory.
+
+  The data directory should have the following structure:
+
+  ```
+  path/to/manual_dir/<dataset_name>/
+    split_name/  # Ex: 'train'
+      label1/  # Ex: 'airplane' or '0015'
+        xxx.png
+        xxy.png
+        xxz.png
+      label2/
+        xxx.png
+        xxy.png
+        xxz.png
+    split_name/  # Ex: 'test'
+      ...
+  ```
+
+  To use it:
+
+  ```
+  builder = tfds.ImageFolder(root_dir='path/to/manual_dir/')
+  print(builder.info)  # Splits, num examples,... are automatically calculated
+  ds = builder.as_dataset(split='split_name', shuffle_files=True)
+  tfds.show_examples(ds, builder.info)
+  ```
+
+  """
+
+  VERSION = core.Version('2.0.0')
+
+  def __init__(self, root_dir):
+    root_dir = os.path.expanduser(root_dir)
+    super(ImageFolder, self).__init__(
+        data_dir=root_dir, version=str(self.VERSION), config=None)
+
+    # Reset `_data_dir` as it should not change to DATA_DIR/Version
+    self._data_dir = root_dir
+
+    # dict[split_name][label_name] = list(img_paths)
+    self._split_label_images, labels = _get_split_label_images(root_dir)
+
+    # Calculate and update DatasetInfo labels
+    self.info.features['label'].names = sorted(list(labels))
+
+    # Calculate and update DatasetInfo splits
+    splits = {
+        split_name: sum(len(l)
+                        for l in self._split_label_images[split_name].values())
+        for split_name in self._split_label_images
+    }
+    split_dict = core.SplitDict(self.name)
+    for split_name, size in splits.items():
+      split_dict.add(
+          core.SplitInfo(name=split_name, shard_lengths=[size]))
+    self.info.update_splits_if_different(split_dict)
+
+  def _info(self):
+    return core.DatasetInfo(
+        builder=self,
+        description='Generic image classification dataset.',
+        features=core.features.FeaturesDict({
+            'image': core.features.Image(),
+            'label': core.features.ClassLabel(),
+            'image/filename': core.features.Text(),
+        }),
+        supervised_keys=('image', 'label'),
+    )
+
+  def _download_and_prepare(self, **kwargs):
+    raise NotImplementedError('No need to call download_and_prepare function.')
+
+  def download_and_prepare(self, **kwargs):
+    raise NotImplementedError('No need to call download_and_prepare function.')
+
+  def _process_ds(self, path, label):
```

Comment:
If the function do not use `self`, it's best to make it a simple function rather than a method. I help the user understand the function do not have side effects & cie.